### PR TITLE
improve authentication conditions and roles mapping

### PR DESCRIPTION
**Pull request by Med1233**
## Description

this one is for updating OpenId authentication form, creating new fields according to API changes

authentification_conditions

roles_mapping

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x (master)

<h2> How this pull request can be tested ? </h2>

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/front_src/src/Authentication/Openid/models.ts [hunk 1]
```diff
@@ -13,40 +13,84 @@ export interface AuthorizationRelationToAPI {
...
+
+export interface Endpoint {
+  customEndpoint: string | null;
+  type: string;
+}
+
+export interface RolesMapping {
+  apply_only_first_role: boolean | null;
+  attribute_path: string;
+  authorization_rules: Array<AuthorizationRelationToAPI> | null;
```
Comment (jeremyjaouen) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  authorization_rules: Array<AuthorizationRelationToAPI>;
```
API will only received or return an array (empty or filled) but will not managed null (will throw an error on back end part)

File: www/front_src/src/Authentication/Openid/models.ts [hunk 2]
```diff
@@ -13,40 +13,84 @@ export interface AuthorizationRelationToAPI {
...
+  endpoint: EndpointToAPI;
+  is_enabled: boolean | null;
+}
+
+export interface AuthConditions {
+  attribute_path: string;
+  authorized_values: Array<string> | null;
```
Comment (jeremyjaouen) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  authorized_values: Array<string>;
```
API will only received or return an array (empty or filled) but will not managed null (will throw an error on back end part)

File: www/front_src/src/Authentication/Openid/models.ts [hunk 3]
```diff
@@ -13,40 +13,84 @@ export interface AuthorizationRelationToAPI {
...
+  type: string;
+}
+
+export interface Endpoint {
+  customEndpoint: string | null;
+  type: string;
+}
+
+export interface RolesMapping {
+  apply_only_first_role: boolean | null;
```
Comment (jeremyjaouen) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  apply_only_first_role: boolean;
```
should always be a boolean

File: www/front_src/src/Authentication/Openid/models.ts [hunk 4]
```diff
@@ -13,40 +13,84 @@ export interface AuthorizationRelationToAPI {
...
+  attribute_path: string;
+  authorization_rules: Array<AuthorizationRelationToAPI> | null;
+  endpoint: EndpointToAPI;
+  is_enabled: boolean | null;
```
Comment (jeremyjaouen) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  is_enabled: boolean;
```
should always be a boolean

File: www/front_src/src/Authentication/Openid/models.ts [hunk 5]
```diff
@@ -13,40 +13,84 @@ export interface AuthorizationRelationToAPI {
...
+}
+
+export interface AuthConditions {
+  attribute_path: string;
+  authorized_values: Array<string> | null;
+  blacklist_client_addresses: Array<string>;
+  endpoint: EndpointToAPI;
+  is_enabled: boolean | null;
```
Comment (jeremyjaouen) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  is_enabled: boolean;
```
should always be a boolean

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Comment by jeremyjaouen** [comment]
Also you should target the feature branch MON-14870 and not develop

File: package.json [hunk 6]
```diff
@@ -109,7 +109,7 @@
     "install": "^0.13.0",
     "jotai": "^1.6.3",
     "json-loader": "^0.5.7",
-    "npm": "^8.12.1",
+    "npm": "^8.19.1",
```
Comment (Thebarda) [inline package.json]: remove this

File: www/front_src/src/Authentication/translatedLabels.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 export const labelActivation = 'Activation';
 export const labelIdentityProvider = 'Identity provider';
-export const labelClientAddresses = 'Client addresses';
+export const labelAuthentificationConditions = 'Authentification conditions';
```
Comment (Thebarda) [inline www/front_src/src/Authentication/translatedLabels.ts]:
```suggestion
export const labelAuthenticationConditions = 'Authentication conditions';
```

File: www/front_src/src/Authentication/Openid/translatedLabels.tsx [hunk 8]
```diff
@@ -55,3 +55,27 @@ export const labelIfWeConsiderTheFollowingExample =
   'If we consider the following example in the JSON received from the authorization endpoint, a relation is defined between the value';
 export const labelTechnical = 'technical';
 export const labelAndAnAccessGroup = 'and an access group:';
+export const labelEnableConditionsOnIdentityProvider =
+  'Enable conditions on identity provider';
+export const labelConditionsAttributePath = 'Conditions attribute path';
+export const labelWhichendpointtheConditionsAttributePathComeFrom =
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/translatedLabels.tsx]:
```suggestion
export const labelWhichEndpointTheConditionsAttributePathComeFrom =
```

File: www/front_src/src/Authentication/Openid/translatedLabels.tsx [hunk 9]
```diff
@@ -55,3 +55,27 @@ export const labelIfWeConsiderTheFollowingExample =
...
+export const labelIntrospectionToken = 'Introspection token';
+export const labelUserIformation = 'User information';
+export const labelOther = 'Other';
+export const labelDefineAuthorizedConditionsValues =
+  'Define authorized conditions values';
+export const labelConditionValue = 'Condition value';
+// roles mapping
+
+export const labelRolesMapping = 'Roles mapping';
+export const labelWhichendpointtheRolesAttributePathComeFrom =
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/translatedLabels.tsx]:
```suggestion
export const labelWhichEndpointTheRolesAttributePathComeFrom =
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 10]
```diff
@@ -33,13 +33,30 @@ import {
   labelDefineRelationAuthorizationValueAndAccessGroup,
   labelDeleteRelation,
   labelAuthorizationKey,
+  labelEnableConditionsOnIdentityProvider,
+  labelConditionsAttributePath,
+  labelWhichendpointtheConditionsAttributePathComeFrom,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
  labelWhichEndpointTheConditionsAttributePathComeFrom,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 11]
```diff
@@ -33,13 +33,30 @@ import {
...
+  labelOther,
+  labelUserIformation,
+  labelIntrospectionToken,
+  labelDefineAuthorizedConditionsValues,
+  labelConditionValue,
+  labelRolesMapping,
+  labelWhichendpointtheRolesAttributePathComeFrom,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
  labelWhichEndpointtheRolesAttributePathComeFrom,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 12]
```diff
@@ -33,13 +33,30 @@ import {
...
+  labelAclAccessGroup,
+  labelDefineYourEndpoint,
 } from '../translatedLabels';
 import { AuthenticationType } from '../models';
 import {
   labelActivation,
   labelAuthorizations,
   labelAutoImportUsers,
-  labelClientAddresses,
+  labelAuthentificationConditions,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
  labelAuthenticationConditions,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 13]
```diff
@@ -88,7 +105,7 @@ export const inputs: Array<InputProps> = [
       options: [],
     },
     fieldName: 'trustedClientAddresses',
-    group: labelClientAddresses,
+    group: labelAuthentificationConditions,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    group: labelAuthenticationConditions,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 14]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
       options: [],
     },
     fieldName: 'blacklistClientAddresses',
-    group: labelClientAddresses,
+    group: labelAuthentificationConditions,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    group: labelAuthenticationConditions,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 15]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+    fieldName: 'conditionsAttributePath',
+    group: labelAuthentificationConditions,
+    label: labelConditionsAttributePath,
+    // required: true,
+    type: InputType.Text,
+  },
+  {
+    fieldName: 'endpointTheConditionsAttributePathComeFrom',
+    group: labelAuthentificationConditions,
+    label: labelWhichendpointtheConditionsAttributePathComeFrom,
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    label: labelWhichEndpointtheConditionsAttributePathComeFrom,
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 16]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+          label: labelIntrospectionToken,
+          value: 'introspection_endpoint',
+        },
+        {
+          label: labelUserIformation,
+          value: 'user_information_endpoint',
+        },
+        {
+          label: labelOther,
+          value: 'custom_endpoint',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]: Use an enum instead for the reusability

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 17]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]: Use an enum instead for the reusability

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 18]
```diff
@@ -59,7 +76,7 @@ const isAuthorizationRelationsFilled = (values: FormikValues): boolean =>
 
 export const inputs: Array<InputProps> = [
   {
-    fieldName: 'isActive',
+    fieldName: 'isEnabled',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'isActive',
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 19]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  authorized_values: Array<string>;
+  blacklist_client_addresses: Array<string>;
+  endpoint: EndpointToAPI;
+  is_enabled: boolean;
+  trusted_client_addresses: Array<string>;
+}
+
 export interface OpenidConfiguration {
   authenticationType: string | null;
+  authentificationConditions: AuthConditions;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  authenticationConditions: AuthConditions;
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 20]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  rolesEndpoint: Endpoint;
+  rolesIsEnabled: boolean;
+  rolesRelations: Array<Relations>;
+}
+
+export interface AuthConditions {
+  blacklistClientAddresses: Array<string>;
+  conditionsAttributePath: string;
+  conditionsAuthorizedValues: Array<string>;
+  enableConditionsOnIdentityProvider: boolean;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  isEnabled: boolean;
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 21]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  rolesApplyOnlyFirstRole: boolean;
+  rolesAttributePath: string;
+  rolesEndpoint: Endpoint;
+  rolesIsEnabled: boolean;
+  rolesRelations: Array<Relations>;
+}
+
+export interface AuthConditions {
+  blacklistClientAddresses: Array<string>;
+  conditionsAttributePath: string;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  attributePath: string;
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 22]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  conditionsAuthorizedValues: Array<string>;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  authorizedValues: Array<string>;
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 23]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  enableConditionsOnIdentityProvider: boolean;
+  endpointTheConditionsAttributePathComeFrom: Endpoint;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  endpoint: Endpoint;
```

File: www/front_src/src/Authentication/Openid/models.ts [hunk 24]
```diff
@@ -13,14 +13,67 @@ export interface AuthorizationRelationToAPI {
...
+  customEndpoint: string;
+  type: string;
+}
+
+export interface RolesMapping {
+  rolesApplyOnlyFirstRole: boolean;
+  rolesAttributePath: string;
+  rolesEndpoint: Endpoint;
+  rolesIsEnabled: boolean;
+  rolesRelations: Array<Relations>;
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/models.ts]:
```suggestion
  applyOnlyFirstRole: boolean;
  attributePath: string;
  endpoint: Endpoint;
  isEnabled: boolean;
  relations: Array<Relations>;
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 25]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
       options: [],
     },
     fieldName: 'blacklistClientAddresses',
-    group: labelClientAddresses,
+    group: labelAuthentificationConditions,
     label: labelBlacklistClientAddresses,
     type: InputType.MultiAutocomplete,
   },
+  {
+    fieldName: 'enableConditionsOnIdentityProvider',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'authenticationConditions.isEnabled',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 26]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+    group: labelAuthentificationConditions,
+    label: labelEnableConditionsOnIdentityProvider,
+    type: InputType.Switch,
+  },
+  {
+    fieldName: 'conditionsAttributePath',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'authenticationConditions.attributePath',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 27]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+    group: labelAuthentificationConditions,
+    label: labelConditionsAttributePath,
+    // required: true,
+    type: InputType.Text,
+  },
+  {
+    fieldName: 'endpointTheConditionsAttributePathComeFrom',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'authenticationConditions.endpoint',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 28]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+          label: labelOther,
+          value: 'custom_endpoint',
+        },
+      ],
+    },
+    type: InputType.Radio,
+    extraTextLabel : labelDefineYourEndpoint
+  },
+  {
+    fieldName: 'conditionsAuthorizedValues',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'authenticationConditions.authorizedValues',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 29]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+        {
+          fieldName: 'conditionValue',
+          label: labelConditionValue,
+          type: InputType.Text,
+        },
+      ],
+      defaultRowValue: {
+        accessGroup: null,
+        claimValue: '',
+      },
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts:238]:
```suggestion
      defaultRowValue: {
        conditionValue: '',
      },
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 30]
```diff
@@ -98,10 +115,68 @@ export const inputs: Array<InputProps> = [
...
+        },
+      ],
+    },
+    type: InputType.Radio,
+    extraTextLabel : labelDefineYourEndpoint
+  },
+  {
+    fieldName: 'conditionsAuthorizedValues',
+    fieldsTable: {
+      additionalFieldsToMemoize: ['contactGroup'],
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 31]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
     label: labelDefineRelationAuthorizationValueAndAccessGroup,
     type: InputType.FieldsTable,
   },
+  {
+    fieldName: 'rolesIsEnabled',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'rolesMapping.isEnabled',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 32]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+    group: labelRolesMapping,
+    label: labelEnableAutoManagement,
+    type: InputType.Switch,
+  },
+  {
+    fieldName: 'rolesApplyOnlyFirstRole',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'rolesMapping.applyOnlyFirstRole',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 33]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+    group: labelRolesMapping,
+    label: labelApplyOnlyFirtsRole,
+    type: InputType.Switch,
+  },
+  {
+    fieldName: 'rolesAttributePath',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'rolesMapping.attributePath',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 34]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+    group: labelRolesMapping,
+    label: labelRolesAttributePath,
+    type: InputType.Text,
+  },
+  {
+    fieldName: 'rolesEndpoint',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'rolesMapping.endpoint',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 35]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+          label: labelOther,
+          value: 'custom_endpoint',
+        },
+      ],
+    },
+    type: InputType.Radio,
+    extraTextLabel : labelDefineYourEndpoint
+  },
+  {
+    fieldName: 'rolesRelations',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
    fieldName: 'rolesMapping.relations',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 36]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+    fieldsTable: {
+      additionalFieldsToMemoize: ['contactGroup'],
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 37]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+      columns: [
+        {
+          fieldName: 'roleValue',
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
          fieldName: 'claimValue',
```

File: www/front_src/src/Authentication/Openid/Form/inputs.ts [hunk 38]
```diff
@@ -277,4 +352,75 @@ export const inputs: Array<InputProps> = [
...
+          },
+          fieldName: 'accessGroup',
+          label: labelAclAccessGroup,
+          type: InputType.SingleConnectedAutocomplete,
+        },
+      ],
+      defaultRowValue: {
+        accessGroup: null,
+        claimValue: '',
+      },
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/Form/inputs.ts]:
```suggestion
      defaultRowValue: {
        accessGroup: null,
        claimValue: '',
      },
```

File: www/front_src/src/Authentication/Openid/useValidationSchema.ts [hunk 39]
```diff
@@ -20,30 +20,57 @@ const useValidationSchema = (): Yup.SchemaOf<OpenidConfiguration> => {
     name: Yup.string().required(t(labelRequired)),
   });
 
-  const authorizationSchema = Yup.object({
+  const rolesRelationSchema = Yup.object({
     accessGroup: namedEntitySchema.nullable().required(t(labelRequired)),
     claimValue: Yup.string().required(t(labelRequired)),
   });
 
+  const groupesRelationSchema = Yup.object({
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/useValidationSchema.ts]:
```suggestion
  const groupsRelationSchema = Yup.object({
```

File: www/front_src/src/Authentication/Openid/useValidationSchema.ts [hunk 40]
```diff
@@ -68,18 +95,26 @@ const useValidationSchema = (): Yup.SchemaOf<OpenidConfiguration> => {
           : schema.nullable();
       },
     ),
+    groupsMapping: Yup.object({
+      attributePath: Yup.string(),
+      endpoint: endpointSchema,
+      isEnabled: switchSchema,
+      relations: Yup.array().of(groupesRelationSchema),
```
Comment (Thebarda) [inline www/front_src/src/Authentication/Openid/useValidationSchema.ts]:
```suggestion
      relations: Yup.array().of(groupsRelationSchema),
```

**Review by Thebarda — APPROVED** [review]

**Review by bdauria — APPROVED** [review]

**Pull request merged**